{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# JM-lab virome pipeline: tutorial\n",
    "This jupyter notebook gives an overview of the commands needed for the primary analysis of raw NGS data with explanations. This is intended as a learning tool for new PhD-students, master students, interns, etc. \n",
    "\n",
    "## Logging in to teaching server\n",
    "For this this tutorial we can work on the teaching server of gbiomed (bmw.gbiomed.kuleuven.be). Everyone with a u- or r-number can connect to this server by ssh-ing to *__'your_r/u-number'@bmw.gbiomed.kuleuven.be__* and giving your intranet password.\n",
    "\n",
    "Perform following actions in the terminal:\n",
    "```bash\n",
    "ssh 'your_r/u-number'@bmw.gbiomed.kuleuven.be\n",
    "```\n",
    "Next, you should give your password connected to your KU Leuven account.\n",
    "\n",
    "## Installing all necessary software\n",
    "### Miniconda\n",
    "(Mini)conda is a package manager from which you can install a lot of (bioinformatics) software. https://docs.conda.io/projects/conda/en/latest/\n",
    "1. Create in your datafolder a new directory and move into that directory:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/home/luna.kuleuven.be/u0140985/data/software\n"
     ]
    }
   ],
   "source": [
    "cd ~/data\n",
    "mkdir software\n",
    "cd software\n",
    "pwd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Download the Miniconda installer with `wget`. Next, run the installation script (`-b` makes the installation run silent and `-p` provides the path where to install Miniconda). When Miniconda is installed, activate the tool by sourcing the initialization script, this simply sets a couple of shell environment variables, and conda command as a shell function. More information: https://docs.conda.io/projects/conda/en/latest/user-guide/install/index.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-01-13 16:49:44--  https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\n",
      "Resolving repo.anaconda.com (repo.anaconda.com)... 104.16.131.3, 104.16.130.3, 2606:4700::6810:8203, ...\n",
      "Connecting to repo.anaconda.com (repo.anaconda.com)|104.16.131.3|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 94235922 (90M) [application/x-sh]\n",
      "Saving to: ‘Miniconda3-latest-Linux-x86_64.sh’\n",
      "\n",
      "Miniconda3-latest-L 100%[===================>]  89.87M  84.1MB/s    in 1.1s    \n",
      "\n",
      "2021-01-13 16:49:45 (84.1 MB/s) - ‘Miniconda3-latest-Linux-x86_64.sh’ saved [94235922/94235922]\n",
      "\n",
      "PREFIX=/home/luna.kuleuven.be/u0140985/data/software/miniconda\n",
      "Unpacking payload ...\n",
      "Collecting package metadata (current_repodata.json): done                       \n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/luna.kuleuven.be/u0140985/data/software/miniconda\n",
      "\n",
      "  added / updated specs:\n",
      "    - _libgcc_mutex==0.1=main\n",
      "    - brotlipy==0.7.0=py38h27cfd23_1003\n",
      "    - ca-certificates==2020.10.14=0\n",
      "    - certifi==2020.6.20=pyhd3eb1b0_3\n",
      "    - cffi==1.14.3=py38h261ae71_2\n",
      "    - chardet==3.0.4=py38h06a4308_1003\n",
      "    - conda-package-handling==1.7.2=py38h03888b9_0\n",
      "    - conda==4.9.2=py38h06a4308_0\n",
      "    - cryptography==3.2.1=py38h3c74f83_1\n",
      "    - idna==2.10=py_0\n",
      "    - ld_impl_linux-64==2.33.1=h53a641e_7\n",
      "    - libedit==3.1.20191231=h14c3975_1\n",
      "    - libffi==3.3=he6710b0_2\n",
      "    - libgcc-ng==9.1.0=hdf63c60_0\n",
      "    - libstdcxx-ng==9.1.0=hdf63c60_0\n",
      "    - ncurses==6.2=he6710b0_1\n",
      "    - openssl==1.1.1h=h7b6447c_0\n",
      "    - pip==20.2.4=py38h06a4308_0\n",
      "    - pycosat==0.6.3=py38h7b6447c_1\n",
      "    - pycparser==2.20=py_2\n",
      "    - pyopenssl==19.1.0=pyhd3eb1b0_1\n",
      "    - pysocks==1.7.1=py38h06a4308_0\n",
      "    - python==3.8.5=h7579374_1\n",
      "    - readline==8.0=h7b6447c_0\n",
      "    - requests==2.24.0=py_0\n",
      "    - ruamel_yaml==0.15.87=py38h7b6447c_1\n",
      "    - setuptools==50.3.1=py38h06a4308_1\n",
      "    - six==1.15.0=py38h06a4308_0\n",
      "    - sqlite==3.33.0=h62c20be_0\n",
      "    - tk==8.6.10=hbc83047_0\n",
      "    - tqdm==4.51.0=pyhd3eb1b0_0\n",
      "    - urllib3==1.25.11=py_0\n",
      "    - wheel==0.35.1=pyhd3eb1b0_0\n",
      "    - xz==5.2.5=h7b6447c_0\n",
      "    - yaml==0.2.5=h7b6447c_0\n",
      "    - zlib==1.2.11=h7b6447c_3\n",
      "\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main\n",
      "  brotlipy           pkgs/main/linux-64::brotlipy-0.7.0-py38h27cfd23_1003\n",
      "  ca-certificates    pkgs/main/linux-64::ca-certificates-2020.10.14-0\n",
      "  certifi            pkgs/main/noarch::certifi-2020.6.20-pyhd3eb1b0_3\n",
      "  cffi               pkgs/main/linux-64::cffi-1.14.3-py38h261ae71_2\n",
      "  chardet            pkgs/main/linux-64::chardet-3.0.4-py38h06a4308_1003\n",
      "  conda              pkgs/main/linux-64::conda-4.9.2-py38h06a4308_0\n",
      "  conda-package-han~ pkgs/main/linux-64::conda-package-handling-1.7.2-py38h03888b9_0\n",
      "  cryptography       pkgs/main/linux-64::cryptography-3.2.1-py38h3c74f83_1\n",
      "  idna               pkgs/main/noarch::idna-2.10-py_0\n",
      "  ld_impl_linux-64   pkgs/main/linux-64::ld_impl_linux-64-2.33.1-h53a641e_7\n",
      "  libedit            pkgs/main/linux-64::libedit-3.1.20191231-h14c3975_1\n",
      "  libffi             pkgs/main/linux-64::libffi-3.3-he6710b0_2\n",
      "  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.1.0-hdf63c60_0\n",
      "  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.1.0-hdf63c60_0\n",
      "  ncurses            pkgs/main/linux-64::ncurses-6.2-he6710b0_1\n",
      "  openssl            pkgs/main/linux-64::openssl-1.1.1h-h7b6447c_0\n",
      "  pip                pkgs/main/linux-64::pip-20.2.4-py38h06a4308_0\n",
      "  pycosat            pkgs/main/linux-64::pycosat-0.6.3-py38h7b6447c_1\n",
      "  pycparser          pkgs/main/noarch::pycparser-2.20-py_2\n",
      "  pyopenssl          pkgs/main/noarch::pyopenssl-19.1.0-pyhd3eb1b0_1\n",
      "  pysocks            pkgs/main/linux-64::pysocks-1.7.1-py38h06a4308_0\n",
      "  python             pkgs/main/linux-64::python-3.8.5-h7579374_1\n",
      "  readline           pkgs/main/linux-64::readline-8.0-h7b6447c_0\n",
      "  requests           pkgs/main/noarch::requests-2.24.0-py_0\n",
      "  ruamel_yaml        pkgs/main/linux-64::ruamel_yaml-0.15.87-py38h7b6447c_1\n",
      "  setuptools         pkgs/main/linux-64::setuptools-50.3.1-py38h06a4308_1\n",
      "  six                pkgs/main/linux-64::six-1.15.0-py38h06a4308_0\n",
      "  sqlite             pkgs/main/linux-64::sqlite-3.33.0-h62c20be_0\n",
      "  tk                 pkgs/main/linux-64::tk-8.6.10-hbc83047_0\n",
      "  tqdm               pkgs/main/noarch::tqdm-4.51.0-pyhd3eb1b0_0\n",
      "  urllib3            pkgs/main/noarch::urllib3-1.25.11-py_0\n",
      "  wheel              pkgs/main/noarch::wheel-0.35.1-pyhd3eb1b0_0\n",
      "  xz                 pkgs/main/linux-64::xz-5.2.5-h7b6447c_0\n",
      "  yaml               pkgs/main/linux-64::yaml-0.2.5-h7b6447c_0\n",
      "  zlib               pkgs/main/linux-64::zlib-1.2.11-h7b6447c_3\n",
      "\n",
      "\n",
      "Preparing transaction: done\n",
      "Executing transaction: done\n",
      "installation finished.\n",
      "(base) no change     /home/luna.kuleuven.be/u0140985/data/software/miniconda/condabin/conda\n",
      "no change     /home/luna.kuleuven.be/u0140985/data/software/miniconda/bin/conda\n",
      "no change     /home/luna.kuleuven.be/u0140985/data/software/miniconda/bin/conda-env\n",
      "no change     /home/luna.kuleuven.be/u0140985/data/software/miniconda/bin/activate\n",
      "no change     /home/luna.kuleuven.be/u0140985/data/software/miniconda/bin/deactivate\n",
      "no change     /home/luna.kuleuven.be/u0140985/data/software/miniconda/etc/profile.d/conda.sh\n",
      "no change     /home/luna.kuleuven.be/u0140985/data/software/miniconda/etc/fish/conf.d/conda.fish\n",
      "no change     /home/luna.kuleuven.be/u0140985/data/software/miniconda/shell/condabin/Conda.psm1\n",
      "no change     /home/luna.kuleuven.be/u0140985/data/software/miniconda/shell/condabin/conda-hook.ps1\n",
      "no change     /home/luna.kuleuven.be/u0140985/data/software/miniconda/lib/python3.8/site-packages/xontrib/conda.xsh\n",
      "no change     /home/luna.kuleuven.be/u0140985/data/software/miniconda/etc/profile.d/conda.csh\n",
      "no change     /home/luna.kuleuven.be/u0140985/.bashrc\n",
      "No action taken.\n",
      "(base) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "wget https://repo.anaconda.com/miniconda/Miniconda3-latest-Linux-x86_64.sh\n",
    "bash Miniconda3-latest-Linux-x86_64.sh -b -p $HOME/data/software/miniconda\n",
    "source $HOME/data/software/miniconda/bin/activate\n",
    "conda init\n",
    "source ~/.bashrc"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. When installing new software with conda the best practice is to create a new conda environment for each part of a project your working on, for example:\n",
    "\n",
    "In this tutorial we will run the virome pipeline, so we will create a conda environment with all software we need to run the pipeline installed in this environment. Then we need to activate this environment to make the software available for use.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/luna.kuleuven.be/u0140985/data/software/miniconda/envs/virome_pipeline\n",
      "\n",
      "  added / updated specs:\n",
      "    - biopython\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    biopython-1.78             |   py38h7b6447c_0         2.1 MB\n",
      "    blas-1.0                   |              mkl           6 KB\n",
      "    ca-certificates-2020.12.8  |       h06a4308_0         121 KB\n",
      "    certifi-2020.12.5          |   py38h06a4308_0         141 KB\n",
      "    intel-openmp-2020.2        |              254         786 KB\n",
      "    mkl-2020.2                 |              256       138.3 MB\n",
      "    mkl-service-2.3.0          |   py38he904b0f_0          62 KB\n",
      "    mkl_fft-1.2.0              |   py38h23d657b_0         157 KB\n",
      "    mkl_random-1.1.1           |   py38h0573a6f_0         341 KB\n",
      "    numpy-1.19.2               |   py38h54aff64_0          22 KB\n",
      "    numpy-base-1.19.2          |   py38hfa32c7d_0         4.2 MB\n",
      "    openssl-1.1.1i             |       h27cfd23_0         2.5 MB\n",
      "    pip-20.3.3                 |   py38h06a4308_0         1.8 MB\n",
      "    setuptools-51.1.2          |   py38h06a4308_4         742 KB\n",
      "    wheel-0.36.2               |     pyhd3eb1b0_0          33 KB\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:       151.3 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main\n",
      "  biopython          pkgs/main/linux-64::biopython-1.78-py38h7b6447c_0\n",
      "  blas               pkgs/main/linux-64::blas-1.0-mkl\n",
      "  ca-certificates    pkgs/main/linux-64::ca-certificates-2020.12.8-h06a4308_0\n",
      "  certifi            pkgs/main/linux-64::certifi-2020.12.5-py38h06a4308_0\n",
      "  intel-openmp       pkgs/main/linux-64::intel-openmp-2020.2-254\n",
      "  ld_impl_linux-64   pkgs/main/linux-64::ld_impl_linux-64-2.33.1-h53a641e_7\n",
      "  libedit            pkgs/main/linux-64::libedit-3.1.20191231-h14c3975_1\n",
      "  libffi             pkgs/main/linux-64::libffi-3.3-he6710b0_2\n",
      "  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.1.0-hdf63c60_0\n",
      "  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.1.0-hdf63c60_0\n",
      "  mkl                pkgs/main/linux-64::mkl-2020.2-256\n",
      "  mkl-service        pkgs/main/linux-64::mkl-service-2.3.0-py38he904b0f_0\n",
      "  mkl_fft            pkgs/main/linux-64::mkl_fft-1.2.0-py38h23d657b_0\n",
      "  mkl_random         pkgs/main/linux-64::mkl_random-1.1.1-py38h0573a6f_0\n",
      "  ncurses            pkgs/main/linux-64::ncurses-6.2-he6710b0_1\n",
      "  numpy              pkgs/main/linux-64::numpy-1.19.2-py38h54aff64_0\n",
      "  numpy-base         pkgs/main/linux-64::numpy-base-1.19.2-py38hfa32c7d_0\n",
      "  openssl            pkgs/main/linux-64::openssl-1.1.1i-h27cfd23_0\n",
      "  pip                pkgs/main/linux-64::pip-20.3.3-py38h06a4308_0\n",
      "  python             pkgs/main/linux-64::python-3.8.5-h7579374_1\n",
      "  readline           pkgs/main/linux-64::readline-8.0-h7b6447c_0\n",
      "  setuptools         pkgs/main/linux-64::setuptools-51.1.2-py38h06a4308_4\n",
      "  six                pkgs/main/linux-64::six-1.15.0-py38h06a4308_0\n",
      "  sqlite             pkgs/main/linux-64::sqlite-3.33.0-h62c20be_0\n",
      "  tk                 pkgs/main/linux-64::tk-8.6.10-hbc83047_0\n",
      "  wheel              pkgs/main/noarch::wheel-0.36.2-pyhd3eb1b0_0\n",
      "  xz                 pkgs/main/linux-64::xz-5.2.5-h7b6447c_0\n",
      "  zlib               pkgs/main/linux-64::zlib-1.2.11-h7b6447c_3\n",
      "\n",
      "\n",
      "\n",
      "Downloading and Extracting Packages\n",
      "biopython-1.78       | 2.1 MB    | ##################################### | 100% \n",
      "intel-openmp-2020.2  | 786 KB    | ##################################### | 100% \n",
      "mkl_fft-1.2.0        | 157 KB    | ##################################### | 100% \n",
      "ca-certificates-2020 | 121 KB    | ##################################### | 100% \n",
      "openssl-1.1.1i       | 2.5 MB    | ##################################### | 100% \n",
      "numpy-1.19.2         | 22 KB     | ##################################### | 100% \n",
      "certifi-2020.12.5    | 141 KB    | ##################################### | 100% \n",
      "mkl_random-1.1.1     | 341 KB    | ##################################### | 100% \n",
      "numpy-base-1.19.2    | 4.2 MB    | ##################################### | 100% \n",
      "setuptools-51.1.2    | 742 KB    | ##################################### | 100% \n",
      "pip-20.3.3           | 1.8 MB    | ##################################### | 100% \n",
      "blas-1.0             | 6 KB      | ##################################### | 100% \n",
      "mkl-service-2.3.0    | 62 KB     | ##################################### | 100% \n",
      "mkl-2020.2           | 138.3 MB  | ##################################### | 100% \n",
      "wheel-0.36.2         | 33 KB     | ##################################### | 100% \n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: done\n",
      "#\n",
      "# To activate this environment, use\n",
      "#\n",
      "#     $ conda activate virome_pipeline\n",
      "#\n",
      "# To deactivate an active environment, use\n",
      "#\n",
      "#     $ conda deactivate\n",
      "\n",
      "(base) (virome_pipeline) Collecting package metadata (current_repodata.json): done\n",
      "Solving environment: failed with initial frozen solve. Retrying with flexible solve.\n",
      "Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.\n",
      "Collecting package metadata (repodata.json): done\n",
      "Solving environment: done\n",
      "\n",
      "## Package Plan ##\n",
      "\n",
      "  environment location: /home/luna.kuleuven.be/u0140985/data/software/miniconda/envs/virome_pipeline\n",
      "\n",
      "  added / updated specs:\n",
      "    - bedtools\n",
      "    - bowtie2\n",
      "    - bwa-mem2\n",
      "    - diamond\n",
      "    - krona\n",
      "    - samtools\n",
      "    - spades\n",
      "    - trimmomatic\n",
      "\n",
      "\n",
      "The following packages will be downloaded:\n",
      "\n",
      "    package                    |            build\n",
      "    ---------------------------|-----------------\n",
      "    bedtools-2.29.2            |       hc088bd4_0        13.8 MB  bioconda\n",
      "    bowtie2-2.4.1              |   py38he513fc3_0        12.6 MB  bioconda\n",
      "    bwa-mem2-2.1               |       he513fc3_0         3.0 MB  bioconda\n",
      "    bzip2-1.0.8                |       h7b6447c_0          78 KB\n",
      "    curl-7.71.1                |       hbc83047_1         140 KB\n",
      "    diamond-0.9.21             |                1         640 KB  bioconda\n",
      "    krb5-1.18.2                |       h173b8e3_0         1.3 MB\n",
      "    krona-2.7.1                |          pl526_3         189 KB  bioconda\n",
      "    libcurl-7.71.1             |       h20c2e04_1         305 KB\n",
      "    libgcc-7.2.0               |       h69d50b8_2         269 KB\n",
      "    libssh2-1.9.0              |       h1ba5d50_1         269 KB\n",
      "    openjdk-8.0.152            |       h7b6447c_3        57.4 MB\n",
      "    perl-5.26.2                |       h14c3975_0        10.5 MB\n",
      "    samtools-1.7               |                1         1.0 MB  bioconda\n",
      "    spades-3.13.1              |                0        13.7 MB  bioconda\n",
      "    tbb-2020.3                 |       hfd86e86_0         1.1 MB\n",
      "    trimmomatic-0.39           |                1         142 KB  bioconda\n",
      "    ------------------------------------------------------------\n",
      "                                           Total:       116.2 MB\n",
      "\n",
      "The following NEW packages will be INSTALLED:\n",
      "\n",
      "  bedtools           bioconda/linux-64::bedtools-2.29.2-hc088bd4_0\n",
      "  bowtie2            bioconda/linux-64::bowtie2-2.4.1-py38he513fc3_0\n",
      "  bwa-mem2           bioconda/linux-64::bwa-mem2-2.1-he513fc3_0\n",
      "  bzip2              pkgs/main/linux-64::bzip2-1.0.8-h7b6447c_0\n",
      "  curl               pkgs/main/linux-64::curl-7.71.1-hbc83047_1\n",
      "  diamond            bioconda/linux-64::diamond-0.9.21-1\n",
      "  krb5               pkgs/main/linux-64::krb5-1.18.2-h173b8e3_0\n",
      "  krona              bioconda/noarch::krona-2.7.1-pl526_3\n",
      "  libcurl            pkgs/main/linux-64::libcurl-7.71.1-h20c2e04_1\n",
      "  libgcc             pkgs/main/linux-64::libgcc-7.2.0-h69d50b8_2\n",
      "  libssh2            pkgs/main/linux-64::libssh2-1.9.0-h1ba5d50_1\n",
      "  openjdk            pkgs/main/linux-64::openjdk-8.0.152-h7b6447c_3\n",
      "  perl               pkgs/main/linux-64::perl-5.26.2-h14c3975_0\n",
      "  samtools           bioconda/linux-64::samtools-1.7-1\n",
      "  spades             bioconda/linux-64::spades-3.13.1-0\n",
      "  tbb                pkgs/main/linux-64::tbb-2020.3-hfd86e86_0\n",
      "  trimmomatic        bioconda/noarch::trimmomatic-0.39-1\n",
      "\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Downloading and Extracting Packages\n",
      "samtools-1.7         | 1.0 MB    | ##################################### | 100% \n",
      "perl-5.26.2          | 10.5 MB   | ##################################### | 100% \n",
      "curl-7.71.1          | 140 KB    | ##################################### | 100% \n",
      "libssh2-1.9.0        | 269 KB    | ##################################### | 100% \n",
      "krb5-1.18.2          | 1.3 MB    | ##################################### | 100% \n",
      "trimmomatic-0.39     | 142 KB    | ##################################### | 100% \n",
      "bwa-mem2-2.1         | 3.0 MB    | ##################################### | 100% \n",
      "libgcc-7.2.0         | 269 KB    | ##################################### | 100% \n",
      "bedtools-2.29.2      | 13.8 MB   | ##################################### | 100% \n",
      "libcurl-7.71.1       | 305 KB    | ##################################### | 100% \n",
      "openjdk-8.0.152      | 57.4 MB   | ##################################### | 100% \n",
      "krona-2.7.1          | 189 KB    | ##################################### | 100% \n",
      "diamond-0.9.21       | 640 KB    | ##################################### | 100% \n",
      "bowtie2-2.4.1        | 12.6 MB   | ##################################### | 100% \n",
      "bzip2-1.0.8          | 78 KB     | ##################################### | 100% \n",
      "tbb-2020.3           | 1.1 MB    | ##################################### | 100% \n",
      "spades-3.13.1        | 13.7 MB   | ##################################### | 100% \n",
      "Preparing transaction: done\n",
      "Verifying transaction: done\n",
      "Executing transaction: | \n",
      "Krona installed.  You still need to manually update the taxonomy\n",
      "databases before Krona can generate taxonomic reports.  The update\n",
      "script is ktUpdateTaxonomy.sh.  The default location for storing\n",
      "taxonomic databases is /home/luna.kuleuven.be/u0140985/data/software/miniconda/envs/virome_pipeline/opt/krona/taxonomy\n",
      "\n",
      "If you would like the taxonomic data stored elsewhere, simply replace\n",
      "this directory with a symlink.  For example:\n",
      "\n",
      "rm -rf /home/luna.kuleuven.be/u0140985/data/software/miniconda/envs/virome_pipeline/opt/krona/taxonomy\n",
      "mkdir /path/on/big/disk/taxonomy\n",
      "ln -s /path/on/big/disk/taxonomy /home/luna.kuleuven.be/u0140985/data/software/miniconda/envs/virome_pipeline/opt/krona/taxonomy\n",
      "ktUpdateTaxonomy.sh\n",
      "\n",
      "\n",
      "done\n",
      "(virome_pipeline) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "conda create -y --name virome_pipeline python\n",
    "conda activate virome_pipeline\n",
    "conda install -y -c bioconda krona samtools bwa-mem2 bowtie2 spades trimmomatic bedtools"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### Downloading taxonomy database for Krona\n",
    "Krona is installed but we still need to run `ktUpdateTaxonomy.sh` to download the taxonomy database, see message below:\n",
    "\n",
    "```sh\n",
    "Krona installed.  You still need to manually update the taxonomy databases before Krona can generate taxonomic reports. The update script is ktUpdateTaxonomy.sh. \n",
    "The default location for storing taxonomic databases is /home/luna.kuleuven.be/u0140985/data/software/miniconda/envs/virome_pipeline/opt/krona/taxonomy\n",
    "\n",
    "If you would like the taxonomic data stored elsewhere, simply replace\n",
    "this directory with a symlink.  For example:\n",
    "\n",
    "rm -rf /home/luna.kuleuven.be/u0140985/data/software/miniconda/envs/virome_pipeline/opt/krona/taxonomy\n",
    "mkdir /path/on/big/disk/taxonomy\n",
    "ln -s /path/on/big/disk/taxonomy /home/luna.kuleuven.be/u0140985/data/software/miniconda/envs/virome_pipeline/opt/krona/taxonomy\n",
    "ktUpdateTaxonomy.sh```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching taxdump.tar.gz...\n",
      "   Fetching checksum...\n",
      "   Checksum for taxdump.tar.gz matches server.\n",
      "Extracting taxonomy...\n",
      "\n",
      "Cleaning up...\n",
      "\n",
      "Finished.\n",
      "\n",
      "(virome_pipeline) "
     ]
    },
    {
     "ename": "",
     "evalue": "1",
     "output_type": "error",
     "traceback": []
    }
   ],
   "source": [
    "ktUpdateTaxonomy.sh"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### From another source\n",
    "Next to Anaconda/Miniconda their are a lot of other possibilities to install software (`pip`, building from source, installing binaries, cloning from github, etc.)\n",
    "\n",
    "As the latest version of Diamond (a sequence aligner for protein and translated DNA searches) is not available through `conda`, we can install it from github by following the instructions (https://github.com/bbuchfink/diamond & installation https://github.com/bbuchfink/diamond/wiki)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "--2021-01-13 23:12:37--  http://github.com/bbuchfink/diamond/releases/download/v2.0.6/diamond-linux64.tar.gz\n",
      "Resolving github.com (github.com)... 140.82.121.4\n",
      "Connecting to github.com (github.com)|140.82.121.4|:80... connected.\n",
      "HTTP request sent, awaiting response... 301 Moved Permanently\n",
      "Location: https://github.com/bbuchfink/diamond/releases/download/v2.0.6/diamond-linux64.tar.gz [following]\n",
      "--2021-01-13 23:12:37--  https://github.com/bbuchfink/diamond/releases/download/v2.0.6/diamond-linux64.tar.gz\n",
      "Connecting to github.com (github.com)|140.82.121.4|:443... connected.\n",
      "HTTP request sent, awaiting response... 302 Found\n",
      "Location: https://github-production-release-asset-2e65be.s3.amazonaws.com/31987083/77690800-3fb3-11eb-8628-ddfbe9c08477?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210113%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210113T221237Z&X-Amz-Expires=300&X-Amz-Signature=e75ab16f5a0f77a206ddf1606483ca7ef8ff4778963a2aa18a08cfed9c55466e&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=31987083&response-content-disposition=attachment%3B%20filename%3Ddiamond-linux64.tar.gz&response-content-type=application%2Foctet-stream [following]\n",
      "--2021-01-13 23:12:37--  https://github-production-release-asset-2e65be.s3.amazonaws.com/31987083/77690800-3fb3-11eb-8628-ddfbe9c08477?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210113%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210113T221237Z&X-Amz-Expires=300&X-Amz-Signature=e75ab16f5a0f77a206ddf1606483ca7ef8ff4778963a2aa18a08cfed9c55466e&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=31987083&response-content-disposition=attachment%3B%20filename%3Ddiamond-linux64.tar.gz&response-content-type=application%2Foctet-stream\n",
      "Resolving github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)... 52.217.70.220\n",
      "Connecting to github-production-release-asset-2e65be.s3.amazonaws.com (github-production-release-asset-2e65be.s3.amazonaws.com)|52.217.70.220|:443... connected.\n",
      "HTTP request sent, awaiting response... 200 OK\n",
      "Length: 4809579 (4.6M) [application/octet-stream]\n",
      "Saving to: ‘diamond-linux64.tar.gz’\n",
      "\n",
      "diamond-linux64.tar 100%[===================>]   4.59M  6.33MB/s    in 0.7s    \n",
      "\n",
      "2021-01-13 23:12:38 (6.33 MB/s) - ‘diamond-linux64.tar.gz’ saved [4809579/4809579]\n",
      "\n"
     ]
    }
   ],
   "source": [
    "cd ~/data/software/\n",
    "mkdir diamond\n",
    "cd diamond\n",
    "wget http://github.com/bbuchfink/diamond/releases/download/v2.0.6/diamond-linux64.tar.gz\n",
    "tar -xzf diamond-linux64.tar.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we still need to put the diamond executable in our `$PATH` so we can call it on the command line from everywhere in the terminal. This can be done by making a `bin` subdirectory in our `software` directory, followed by creating a symlink from the `diamond` executable to this bin directory and finally export `bin` to our `$PATH` by adding it to your `.profile` or `.bash_profile` file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd ~/data/software\n",
    "mkdir bin\n",
    "cd bin/\n",
    "ln -s ~/data/software/diamond/diamond ."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, open the `.profile` file with `nano` and add following line to the bottom of the file:\n",
    "```bash\n",
    "PATH=\"$PATH:~/data/software/bin\"\n",
    "```\n",
    "More documentation on where and how to set the `PATH` variable in these two topics: \n",
    "* https://superuser.com/questions/183870/difference-between-bashrc-and-bash-profile/183980#183980 \n",
    "* https://unix.stackexchange.com/questions/26047/how-to-correctly-add-a-path-to-path\n",
    "\n",
    "\n",
    "When you `source` your `.profile` file, you should now be able to call diamond.\n",
    "```bash\n",
    "source ~/.profile\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "diamond v2.0.6.144 (C) Max Planck Society for the Advancement of Science\n",
      "Documentation, support and updates available at http://www.diamondsearch.org\n",
      "\n",
      "Syntax: diamond COMMAND [OPTIONS]\n",
      "\n",
      "Commands:\n",
      "makedb\tBuild DIAMOND database from a FASTA file\n",
      "blastp\tAlign amino acid query sequences against a protein reference database\n",
      "blastx\tAlign DNA query sequences against a protein reference database\n",
      "view\tView DIAMOND alignment archive (DAA) formatted file\n",
      "help\tProduce help message\n",
      "version\tDisplay version information\n",
      "getseq\tRetrieve sequences from a DIAMOND database file\n",
      "dbinfo\tPrint information about a DIAMOND database file\n",
      "test\tRun regression tests\n",
      "\n",
      "General options:\n",
      "--threads (-p)           number of CPU threads\n",
      "--db (-d)                database file\n",
      "--out (-o)               output file\n",
      "--outfmt (-f)            output format\n",
      "\t0   = BLAST pairwise\n",
      "\t5   = BLAST XML\n",
      "\t6   = BLAST tabular\n",
      "\t100 = DIAMOND alignment archive (DAA)\n",
      "\t101 = SAM\n",
      "\n",
      "\tValue 6 may be followed by a space-separated list of these keywords:\n",
      "\n",
      "\tqseqid means Query Seq - id\n",
      "\tqlen means Query sequence length\n",
      "\tsseqid means Subject Seq - id\n",
      "\tsallseqid means All subject Seq - id(s), separated by a ';'\n",
      "\tslen means Subject sequence length\n",
      "\tqstart means Start of alignment in query\n",
      "\tqend means End of alignment in query\n",
      "\tsstart means Start of alignment in subject\n",
      "\tsend means End of alignment in subject\n",
      "\tqseq means Aligned part of query sequence\n",
      "\tfull_qseq means Query sequence\n",
      "\tsseq means Aligned part of subject sequence\n",
      "\tfull_sseq means Subject sequence\n",
      "\tevalue means Expect value\n",
      "\tbitscore means Bit score\n",
      "\tscore means Raw score\n",
      "\tlength means Alignment length\n",
      "\tpident means Percentage of identical matches\n",
      "\tnident means Number of identical matches\n",
      "\tmismatch means Number of mismatches\n",
      "\tpositive means Number of positive - scoring matches\n",
      "\tgapopen means Number of gap openings\n",
      "\tgaps means Total number of gaps\n",
      "\tppos means Percentage of positive - scoring matches\n",
      "\tqframe means Query frame\n",
      "\tbtop means Blast traceback operations(BTOP)\n",
      "\tcigar means CIGAR string\n",
      "\tstaxids means unique Subject Taxonomy ID(s), separated by a ';' (in numerical order)\n",
      "\tsscinames means unique Subject Scientific Name(s), separated by a ';'\n",
      "\tsskingdoms means unique Subject Super Kingdom(s), separated by a ';'\n",
      "\tskingdoms means unique Subject Kingdom(s), separated by a ';'\n",
      "\tsphylums means unique Subject Phylum(s), separated by a ';'\n",
      "\tstitle means Subject Title\n",
      "\tsalltitles means All Subject Title(s), separated by a '<>'\n",
      "\tqcovhsp means Query Coverage Per HSP\n",
      "\tscovhsp means Subject Coverage Per HSP\n",
      "\tqtitle means Query title\n",
      "\tqqual means Query quality values for the aligned part of the query\n",
      "\tfull_qqual means Query quality values\n",
      "\tqstrand means Query strand\n",
      "\n",
      "\tDefault: qseqid sseqid pident length mismatch gapopen qstart qend sstart send evalue bitscore\n",
      "--verbose (-v)           verbose console output\n",
      "--log                    enable debug log\n",
      "--quiet                  disable console output\n",
      "--header                 Write header lines to blast tabular format.\n",
      "\n",
      "Makedb options:\n",
      "--in                     input reference file in FASTA format\n",
      "--taxonmap               protein accession to taxid mapping file\n",
      "--taxonnodes             taxonomy nodes.dmp from NCBI\n",
      "--taxonnames             taxonomy names.dmp from NCBI\n",
      "\n",
      "Aligner options:\n",
      "--query (-q)             input query file\n",
      "--strand                 query strands to search (both/minus/plus)\n",
      "--un                     file for unaligned queries\n",
      "--al                     file or aligned queries\n",
      "--unfmt                  format of unaligned query file (fasta/fastq)\n",
      "--alfmt                  format of aligned query file (fasta/fastq)\n",
      "--unal                   report unaligned queries (0=no, 1=yes)\n",
      "--max-target-seqs (-k)   maximum number of target sequences to report alignments for (default=25)\n",
      "--top                    report alignments within this percentage range of top alignment score (overrides --max-target-seqs)\n",
      "--max-hsps               maximum number of HSPs per target sequence to report for each query (default=1)\n",
      "--range-culling          restrict hit culling to overlapping query ranges\n",
      "--compress               compression for output files (0=none, 1=gzip)\n",
      "--evalue (-e)            maximum e-value to report alignments (default=0.001)\n",
      "--min-score              minimum bit score to report alignments (overrides e-value setting)\n",
      "--id                     minimum identity% to report an alignment\n",
      "--query-cover            minimum query cover% to report an alignment\n",
      "--subject-cover          minimum subject cover% to report an alignment\n",
      "--mid-sensitive          enable mid-sensitive mode (default: fast)\n",
      "--sensitive              enable sensitive mode (default: fast)\n",
      "--more-sensitive         enable more sensitive mode (default: fast)\n",
      "--very-sensitive         enable very sensitive mode (default: fast)\n",
      "--ultra-sensitive        enable ultra sensitive mode (default: fast)\n",
      "--block-size (-b)        sequence block size in billions of letters (default=2.0)\n",
      "--index-chunks (-c)      number of chunks for index processing (default=4)\n",
      "--tmpdir (-t)            directory for temporary files\n",
      "--parallel-tmpdir        directory for temporary files used by multiprocessing\n",
      "--gapopen                gap open penalty\n",
      "--gapextend              gap extension penalty\n",
      "--frameshift (-F)        frame shift penalty (default=disabled)\n",
      "--long-reads             short for --range-culling --top 10 -F 15\n",
      "--matrix                 score matrix for protein alignment (default=BLOSUM62)\n",
      "--custom-matrix          file containing custom scoring matrix\n",
      "--comp-based-stats       composition based statistics mode (0-4)\n",
      "--masking                enable tantan masking of repeat regions (0/1=default)\n",
      "--query-gencode          genetic code to use to translate query (see user manual)\n",
      "--salltitles             include full subject titles in DAA file\n",
      "--sallseqid              include all subject ids in DAA file\n",
      "--no-self-hits           suppress reporting of identical self hits\n",
      "--taxonlist              restrict search to list of taxon ids (comma-separated)\n",
      "--taxon-exclude          exclude list of taxon ids (comma-separated)\n",
      "\n",
      "Advanced options:\n",
      "--algo                   Seed search algorithm (0=double-indexed/1=query-indexed)\n",
      "--bin                    number of query bins for seed search\n",
      "--min-orf (-l)           ignore translated sequences without an open reading frame of at least this length\n",
      "--freq-sd                number of standard deviations for ignoring frequent seeds\n",
      "--id2                    minimum number of identities for stage 1 hit\n",
      "--xdrop (-x)             xdrop for ungapped alignment\n",
      "--band                   band for dynamic programming computation\n",
      "--shapes (-s)            number of seed shapes (default=all available)\n",
      "--shape-mask             seed shapes\n",
      "--multiprocessing        enable distributed-memory parallel processing\n",
      "--mp-init                initialize multiprocessing run\n",
      "--ext-chunk-size         chunk size for adaptive ranking (default=auto)\n",
      "--no-ranking             disable ranking heuristic\n",
      "--ext                    Extension mode (banded-fast/banded-slow/full)\n",
      "--culling-overlap        minimum range overlap with higher scoring hit to delete a hit (default=50%)\n",
      "--taxon-k                maximum number of targets to report per species\n",
      "--range-cover            percentage of query range to be covered for range culling (default=50%)\n",
      "--dbsize                 effective database size (in letters)\n",
      "--no-auto-append         disable auto appending of DAA and DMND file extensions\n",
      "--xml-blord-format       Use gnl|BL_ORD_ID| style format in XML output\n",
      "--stop-match-score       Set the match score of stop codons against each other.\n",
      "--tantan-minMaskProb     minimum repeat probability for masking (default=0.9)\n",
      "--file-buffer-size       file buffer size in bytes (default=67108864)\n",
      "--memory-limit (-M)      Memory limit for extension stage in GB\n",
      "--no-unlink              Do not unlink temporary files.\n",
      "\n",
      "View options:\n",
      "--daa (-a)               DIAMOND alignment archive (DAA) file\n",
      "--forwardonly            only show alignments of forward strand\n",
      "\n",
      "Getseq options:\n",
      "--seq                    Sequence numbers to display.\n",
      "\n",
      "Online documentation at http://www.diamondsearch.org\n"
     ]
    }
   ],
   "source": [
    "diamond help"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```bash\n",
    "ls | cut -f1 -d '_'| sort -u > names.txt\n",
    "while read line; do cat ${line}_*_*_R1_*.fastq.gz > $line.R1.fastq.gz; done < names.txt\n",
    "while read line; do cat ${line}_*_*_R2_*.fastq.gz > $line.R2.fastq.gz; done < names.txt\n",
    "rm *L00*```\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Bash",
   "language": "bash",
   "name": "bash"
  },
  "language_info": {
   "codemirror_mode": "shell",
   "file_extension": ".sh",
   "mimetype": "text/x-sh",
   "name": "bash"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
